accepted/0000-sbom-command.md

```diff
@@ -0,0 +1,461 @@
+
+`--package-lock-only` - Constructs the SBOM based on the tree described by the _package-lock.json_, rather than the contents of _node_modules_. Defaults to _false_. If the _node_modules_ folder is not present, this flag will be required in order to generate an SBOM.
+ 
+`--workspace` - Runs the command only in the context of the specified workspace (can be set multiple times).
+
+`--workspaces` - Runs the command in the context of all configured workspaces. 
+
+If the user runs the `sbom` command without first installing the dependencies for the project (i.e. there is no _node_modules_ folder present) an error will be displayed. An SBOM can be generated solely based on the contents of the _package-lock.json_ but requires the user to explicitly specify the `--package-lock-only` flag.
+
+Initially, we'll support the most widely used versions of the SPDX and CycloneDX specifications (likely v2.3 for SPDX and v1.4 for CycloneDX). Best effort will be made to support new versions as they gain adoption across the ecosystem.
+
+
+## Rationale and Alternatives
+
+There are a few existing tools which can be used to generate an SBOM from an npm project:
+
+* <code>[@cyclonedex/cyclonedx-npm](https://www.npmjs.com/package/@cyclonedx/cyclonedx-npm)</code> - A CLI for generating a CycloneDX-style SBOM from an npm project. This project is written in TypeScript and actually invokes <code>npm-ls</code> in order to get dependency information for the project.
+* <code>[spdx-sbom-generator](https://github.com/opensbom-generator/spdx-sbom-generator)</code> - A CLI for generating SPDX-style SBOMs for a number of different package managers (including npm). Currently, this tool only works with npm projects using <em>lockfileVersion</em> 1 so it’s not viable for a large number of projects (current <em>lockfileVersion</em> is 3)
+
+While you can effectively generate the same output we’re proposing with this combination of tools, there is value in having this capability supported directly in npm. Beyond the obvious developer-experience benefit of having SBOM generation baked-in to the CLI, it gives us a future path to do things like automatic-signing of SBOMs or integration of SBOMs into the package publishing process.
+
+
+## Implementation
+
+The `npm-sbom` command is similar in function to `npm-ls` command and will likely utilize a similar implementation. We’ll use <code>[arborist](https://github.com/npm/cli/tree/latest/workspaces/arborist)</code> to construct the dependency tree and the <code>[treeverse](https://github.com/isaacs/treeverse)</code> library to traverse the tree and assemble the SBOM.
+
+
+### Local (Linked) Dependencies
+
+The generated SBOM will exclude any dependencies which are referenced via links on the local file system. A warning will be displayed for any linked dependency informing the user that it was omitted from the SBOM.
```

Comment:
For which reasons shall these dependencies be omitted?
Hiding these would drop important information.

Reply:
My thinking was that an SBOM which references linked packages ends-up being very specific to the environment in which it is generated -- it would be difficult for anyone else to reason about or reproduce the SBOM contents.

That said, there may well be use cases where the user simply wants a strict representation of the project's dependencies even if some of those are links. Perhaps a flag that would include links? Or maybe the links are included by default but you still get a warning message in the output?

Reply:
before I go into details, let me ask the following: 
does `npm ls` list information about locally linked packages?

Reply:
for the same reasons `npm ls` lists these information, I'd argue they must be part of an SBOM.

> an SBOM which references linked packages ends-up being very specific to the environment in which it is generated

One would want to have that specific SBOM for the specific setup, all dependencies listed in my SBOM, regardless whether they are linked or not, so one can do proper risk assessment. 

> [...] it would be difficult for anyone else to reason about [...]

why would that be? Could you explain further? 

> [...] or reproduce the SBOM contents.

As long as I can reproduce the SBOM in the same specific environment, all is good.

here are example setups, so we have a common ground to discuss the meta:
- https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/demo/local-workspaces
- https://github.com/CycloneDX/cyclonedx-node-npm/tree/main/demo/local-dependencies